src/pip/_internal/utils/misc.py

```diff
@@ -953,6 +962,32 @@ def _make_build_dir(build_dir):
     write_delete_marker_file(build_dir)
 
 
+class HiddenText(str):
+
+    def __repr__(self):
+        return '<HiddenText {!r}>'.format(str(self))
+
+    def __str__(self):
+        return self.redacted
+
+    @property
+    def raw(self):
+        return str.__str__(self)
+
+
+def hide_url(raw):
```

Comment:
It would be good to add unit tests of this function (and the other) testing some basic properties of the return value (e.g. `repr()`, `str()`, the attribute access, etc.).